### ISSUE TITLE
Use non-deterministic precision for float operations that do not have guaranteed precision

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/miri#3555]
As per https://github.com/rust-lang/rust/pull/124609, these operations behave non-deterministically, so we should implement that in Miri. The `apply_random_float_error` function (currently only used for x86 intrinsics) could be used to do that.

**Comment by RalfJung** [comment]
This should also be used for fast-math and "algebraic" float operations.

**Comment by LorrensP-2158466** [comment]
Hi, i would like to try this. I have 2 questions though. Do you mean that we should add the float_error to this part of the code?

https://github.com/rust-lang/miri/blob/95558badc942353582b742b30950412d46d69a06/src/intrinsics/mod.rs#L247C1-L263C18

And should we use the same magnitude as the x86 instrinics? https://github.com/rust-lang/miri/blob/95558badc942353582b742b30950412d46d69a06/src/shims/x86/mod.rs#L413

**Comment by RalfJung** [comment]
Yes, all of those there except for sqrt should apply a float error.

There's no fixed value for the magnitude of the error so we might as well pick the same as the x86 intrinsic.

**Comment by LorrensP-2158466** [comment]
Hi, I think I successfully implemented it for all but 2 match arms because I don't know how to do it nicely. It is about the `f*_algebraic` operations:
```rust
  | "fadd_algebraic"
  | "fsub_algebraic"
  | "fmul_algebraic"
  | "fdiv_algebraic"
  | "frem_algebraic"
  => {
      let [a, b] = check_arg_count(args)?;
      let a = this.read_immediate(a)?;
      let b = this.read_immediate(b)?;
      let op = match intrinsic_name {
          "fadd_algebraic" => mir::BinOp::Add,
          "fsub_algebraic" => mir::BinOp::Sub,
          "fmul_algebraic" => mir::BinOp::Mul,
          "fdiv_algebraic" => mir::BinOp::Div,
          "frem_algebraic" => mir::BinOp::Rem,
          _ => bug!(),
      };
      let res = this.binary_op(op, &a, &b)?;
      // `binary_op` already called `generate_nan` if necessary.
      this.write_immediate(*res, dest)?;
  }
```
I don't know how I can nicely implement this.
I thought about converting to a scalar int `to_scalar_int()` and then matching on the float type converting to that type, applying the error, and then writing the scalar. But maybe there is a better way?

And the `f*_fast` operation. It says this in the code:
https://github.com/rust-lang/miri/blob/c2a376121b7756c73b250bdba82f92d8a47c25a4/src/intrinsics/mod.rs#L424C1-L426C51

Should I follow the second comment from RalfJung or this comment in the code?

Also, is this the conventional way of asking for help? Or is Zullip used more? Or another way?

Thanks for the help!

**Comment by RalfJung** [comment]
I don't think I quite understand what the problem is. I would propose you open a PR that implements this for one or two (blocks of) shims, so that we can give feedback on a concrete implementation, and hopefully that will help clarify these things as well.

Regarding asking for help, [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/269128-miri) is a good place for that. :)

> Should I follow the second comment from RalfJung or this comment in the code?

Note that this refers to different non-determinism than what the issue is about. That comment is about making the NaN payload bits and sign non-deterministic.

**Comment by LorrensP-2158466** [comment]
I looked through the code and besides `pow` I also noticed `ldexp` (or the `equivalent`) is missing. I remember not being sure what the precision guarantees were and being busy with the ULP approx test I forgot to ask, sorry :(.

Everything else get's the 16ULP error.

**Comment by LorrensP-2158466** [comment]
So the list is:
- In shims/foreign_items.rs:
```
cbrtf - cbrt (f32/f64)
coshf - cosh (f32/f64)
sinhf - sinh (f32/f64)
tanf - tan (f32/f64)
tanhf - tanh (f32/f64)
acosf - acos (f32/f64)
asinf - asin (f32/f64)
atanf - atan (f32/f64)
log1pf - log1p (f32/f64)
expm1f - expm1 (f32/f64)
tgammaf - tgamma (f32/f64)
erff - erf (f32/f64)
erfcf - erfc (f32/f64)
_hypotf - _hypot (f32/f64)
hypotf - hypot (f32/f64)
atan2f - atan2 (f32/f64)
fdimf - fdim (f32/f64)
lgammaf_r - lgamma_r (f32/f64)

f64 only, they are together in 1 block because they are the same, unsure if precision is infinite or not:
_ldexp)
ldexp
scalbn
```

-  ~~In `intrinsics/mod.rs`, we still have the fused multiply-add intrinsic.~~ These guarantee precision.
- There are also floating-point operations in `shims/simd.rs`, but I'm unsure if this should be handle atm.

**Comment by RalfJung** [comment]
The FMA operation has guaranteed precision.

SIMD operations have guaranteed precision if their scalar variant does... but fsin, fcos and friends should indeed also become non-deterministic.

**Comment by LorrensP-2158466** [comment]
May I do this in 1 pr?

**Comment by RalfJung** [comment]
Please keep the SIMD operations in a separate PR.

**Comment by AaronKutch** [comment]
I'm not sure if this should be on by default. Rounding even a little in the wrong way over a boundary can lead to bad things. My crate tests were failing when I tested them today for the first time in a while and discovered that there are mismatches in constants critical to constant serialization. Reduced case for an instance from my `awint` crate:
```
use core::ops::Mul;

fn main() {
    let a = 4usize as f64;
    let b = a.log2();
    let c = b.powi(-1);
    let d = c.mul((1 << 15) as f64);
    let e = d.ceil();
    let f = e as u16;
    let equiv = (4usize as f64).log2().powi(-1).mul((1 << 15) as f64).ceil() as u16;
    println!("{a} {b} {c} {d} {e} {f} {equiv}");

    assert_eq!(
        16384,
        (4usize as f64).log2().powi(-1).mul((1 << 15) as f64).ceil() as u16
    );
}
```
This will print nice values like
```
4 2 0.5 16384 16384 16384 16384
```
when working correctly. When commenting and uncommenting things to influence the determinism, Miri can produce stuff like
```
4 1.9999999999999982 0.5000000000000007 16384.000000000022 16385 16385 16384
```
And the assertion can decide to fail or not fail some times.

 It happens that the current implementation uses hardcoded constants and only checks against them in a test. In another situation I could have used const-eval to arrive at these constants, in which case they would be plain wrong and inconsistent across builds.

Edit: should I file this in a separate issue?

Edit: I should explain that the reason my own methods are not insane, is that I am using rounding operations in the more conservative direction, and the input values are tiny compared to `f64` precision, the real issue is that just blindly applying non-determinism can lead to otherwise impossible criticalities being crossed. Or maybe this is just a really weird edge case that I should just accept and apply `#[cfg(not(miri))]` for, it warns users about potential non-determinism across architectures in their code.

**Comment by RalfJung** [comment]
@AaronKutch thanks for the feedback! The docs for those functions are fairly clear that they don't provide guaranteed precision. And rounding with `ceil` when the input is near an integer obviously has a big non-continuity, which I guess is what you are hitting here. The code needs more explanation for others to make sense of it, so I can't tell which exact result you are objecting to.

So... is it that you think the functions actually ought to be documented as giving more precise results? Or is your point that you'd prefer to ignore the docs and rely on undocumented precision and hope this won't break in future versions of Rust? Or a third option I haven't understood?

FWIW we do have a flag for people that don't want to find bugs related to floating-point non-determinism: `-Zmiri-deterministic-floats`. Currently this flags disables all float non-det, including the one for NaNs and "opportunistically fused multiply-add". We could offer something more fine-grained that just removes the "random noise" for imprecise operations but keeps NaN non-determinism.

EDIT: https://github.com/rust-lang/miri/pull/4480 implements such a flag, `-Zmiri-no-extra-rounding-error`. It should let you avoid `cfg(miri)`, but I am still curious about the details of your use-case.

**Comment by LorrensP-2158466** [comment]
Now we could also easily add non-determinism to the SIMD operations, especially with your pr @RalfJung :
https://github.com/rust-lang/miri/blob/9396463e426575026ba3c73957c766c5545e30b3/src/intrinsics/simd.rs#L116-L128

Is it okay if I do this as well?

**Comment by RalfJung** [comment]
Please wait until <https://github.com/rust-lang/miri/pull/4484> lands. But yeah then feel free to give it a try :)

**Comment by LorrensP-2158466** [comment]
I was a little too confident with the rust pr.^^
